### PR TITLE
Add new `RSpec/VariableDefinition` cop

### LINKED ISSUE #908: Cop Idea: Ensure let/subject are used with symbols not strings
And provide an option to use strings.

```ruby
# bad
let('first_name') { 'Adam' }
subject('user') { create_user }

#good
let(:first_name) { 'Adam' }
subject(:user) { create_user } 
```

This was brought up by @Darhazer in [this comment](https://github.com/rubocop-hq/rubocop-rspec/pull/907#discussion_r418586020)

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #908

Also refactor common matcher into a module & change `RSpec/VariableName` cop to consider interpolated strings & symbols (they are parsed as `dsym` & `dstr` nodes).

---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Updated documentation.
* [x] Added an entry to the [changelog](https://github.com/rubocop-hq/rubocop-rspec/blob/master/CHANGELOG.md) if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have created a new cop:

* [x] Added the new cop to `config/default.yml`.
* [x] The cop documents examples of good and bad code.
* [x] The tests assert both that bad code is reported and that good code is not reported.

### CODE REVIEW

**Comment by tejasbubane** [comment]
I made some changes to `RSpec/VariableName` cop in the process, let me know if you prefer them in a separate commit.

File: lib/rubocop/cop/rspec/variable_definition.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks that variables are defined using symbols or strings.
+      #
+      # @example EnforcedStyle: symbols (default)
+      #   # bad
+      #   let("user_name") { 'Adam' }
+      #   subject('user') { create_user }
```
Comment (pirj) [inline lib/rubocop/cop/rspec/variable_definition.rb:11]: Double vs single quote - let's use one of those consistently.

File: spec/rubocop/cop/rspec/variable_definition_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    it 'registers an offense for string name' do
+      expect_offense(<<~RUBY)
+        let("user_name") { 'Adam' }
+            ^^^^^^^^^^^ Use symbols for variable names.
+      RUBY
+    end
+
+    it 'registers an offense for interpolated string' do
+      expect_offense(<<~'RUBY')
```
Comment (pirj) [inline spec/rubocop/cop/rspec/variable_definition_spec.rb:17]: 👍 for using `'RUBY'`

**Comment by pirj** [comment]
Triple C-C-C-COMBO! 🤜 🤜 🤜

**Comment by tejasbubane** [comment]
@pirj Changes done.

File: lib/rubocop/cop/rspec/variable_definition.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks that variables are defined using symbols or strings.
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/variable_definition.rb:6]: Reading this description in our documentation, I would think “do they mean local variables or instance variables?”. It’s not immediately obvious that by “variables” we mean the ones you define by calling `let`, `subject`, etc.

Comment (Darhazer) [reply lib/rubocop/cop/rspec/variable_definition.rb:6]: I agree that we need better wording. Strictly speaking, it's a memorized helper method and not a variable

**Comment by pirj** [comment]
@tejasbubane 
```
$ g p --force tejasbubane
...
remote: Resolving deltas: 100% (13/13), completed with 13 local objects.
To github.com:tejasbubane/rubocop-rspec.git
 ! [remote rejected] variable-definition-cop -> variable-definition-cop (permission denied)
```
Can you please check if you have "allow edits from maintainers" checkbox checked in this pull request?

My patch to improve cop descriptions:
```diff
1bc049f 2020-05-09 Phil Pirozhkov    (origin/variable-definition-cop) fixup! Address code review 

diff --git a/config/default.yml b/config/default.yml
index fe8c6f9..21e4d52 100644
--- a/config/default.yml
+++ b/config/default.yml
@@ -455,7 +455,7 @@ RSpec/UnspecifiedException:
   StyleGuide: https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/UnspecifiedException
 
 RSpec/VariableDefinition:
-  Description: Checks that variables are defined using symbols or strings.
+  Description: Checks that memoized helpers names are symbols or strings.
   Enabled: true
   EnforcedStyle: symbols
   SupportedStyles:
@@ -465,7 +465,7 @@ RSpec/VariableDefinition:
   StyleGuide: https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/VariableDefinition
 
 RSpec/VariableName:
-  Description: This cop makes sure that all variables use the configured style.
+  Description: Checks that memoized helper names use the configured style.
   Enabled: true
   EnforcedStyle: snake_case
   SupportedStyles:
diff --git a/lib/rubocop/cop/rspec/variable_definition.rb b/lib/rubocop/cop/rspec/variable_definition.rb
index 16b7b50..5981d07 100644
--- a/lib/rubocop/cop/rspec/variable_definition.rb
+++ b/lib/rubocop/cop/rspec/variable_definition.rb
@@ -3,7 +3,7 @@
 module RuboCop
   module Cop
     module RSpec
-      # Checks that variables are defined using symbols or strings.
+      # Checks that memoized helpers names are symbols or strings.
       #
       # @example EnforcedStyle: symbols (default)
       #   # bad
diff --git a/lib/rubocop/cop/rspec/variable_name.rb b/lib/rubocop/cop/rspec/variable_name.rb
index 463eefb..7fa2228 100644
--- a/lib/rubocop/cop/rspec/variable_name.rb
+++ b/lib/rubocop/cop/rspec/variable_name.rb
@@ -3,7 +3,7 @@
 module RuboCop
   module Cop
     module RSpec
-      # This cop makes sure that all variables use the configured style.
+      # Checks that memoized helper names use the configured style.
       #
       # @example EnforcedStyle: snake_case (default)
       #   # bad
```

**Comment by tejasbubane** [comment]
@pirj I have ticked the checkbox. You should be able to push now.

**Comment by tejasbubane** [comment]
OK seems like github does not persist the checked box (refreshing the page it is unchecked again). I'm going to make these changes and push.

**Comment by pirj** [comment]
Thank you for yet another contribution!

File: spec/rubocop/cop/rspec/variable_definition_spec.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        let("user-#{id}") { 'Adam' }
```
Comment (bquorning) [inline spec/rubocop/cop/rspec/variable_definition_spec.rb:18]: Sorry for adding a comment to a long closed PR, but… what’s the benefit of converting this interpolated string into a symbol? Sure, we’re consistently passing the same _type_ to our `let` calls, but it feels like double work to not just pass the interpolated string.

Comment (pirj) [reply spec/rubocop/cop/rspec/variable_definition_spec.rb:18]: Good point.

It was originally brought up here https://github.com/rubocop-hq/rubocop-rspec/pull/907#discussion_r418586020 by @Darhazer 

Interpolated variable names seem to be an exception (though I remember someone mentioned it was used in some specific specs).

I'm all for adding an exception for interpolated var names (just not up for it due to my overflown queue of things to fix).

**Pull request merged**